src/lib/FreeC/IR/Subterm.hs

```diff
@@ -261,6 +263,13 @@ findSubterms :: Subterm a => (a -> Bool) -> a -> [a]
 findSubterms predicate term =
   filter predicate (map (fromJust . selectSubterm term) (allPos term))
 
+-- | Gets the first subterm of the given expression that satisfies the
+--   provided predicate.
+--
+--   Return @Nothing@ if there is no such subterm.
+findFirstSubterm :: Subterm a => (a -> Bool) -> a -> Maybe a
+findFirstSubterm = listToMaybe .: findSubterms
+
```

Comment:
After reading this definition, I was wondering why the used function `findSubterms` uses `fromJust` rather than `catMaybes` to get rid of `Nothing`s afterwards. Do you really need the run-time error of `fromJust` to be triggered here? Yes, it's safe, because all positions of `term` should yield a valid sub term, but still I'm not sure this is the best way to go here.

Additionally, I was wondering if using [`First`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Semigroup.html#t:First) isn't the usual way to define a function like `findFirstSubTerm`, but perhaps it only overcomplicates the code -- not sure about that.

Reply:
I think the primary reason for using `fromJust` was consistency with `findSubtermPos` which cannot be expressed using `catMaybes`. The potential run-time error could actually be beneficial in this case. It allows us to spot mistakes in the implementation while `catMaybes` would silently ignore such bugs.

A better solution could be to add an alternative to `selectSubterm` that does not return a `Maybe` but prints are more useful error message than `fromJust` by mentioning the position of the subterm that could not be found. Such a function would be also useful since the pattern of removing `Maybe` from the result of `seletSubterm` can be found in [other parts of the code](https://github.com/FreeProving/free-compiler/blob/master/src/lib/FreeC/Backend/Coq/Converter/FuncDecl/Rec/WithHelpers.hs#L151) as well.

---

I'm not sure how you would use `First` in this case. It would be possible to replace `listToMaybe` with concatenation over the `Maybe (First a)` monoid. But as you've said already, that would just over complicate the code.

```haskell
import           Data.Semigroup                 ( First(..) )

findFirstSubterm :: Subterm a => (a -> Bool) -> a -> Maybe a
findFirstSubterm = fmap getFirst . mconcat . map (Just . First) .: findSubterms
```

Reply:
Should we add an issue for the modification of `selectSubterm` with a custom error message?

Reply:
I've added an enhancement proposal #52.